src/exe/udp_client_test/main.cpp

```diff
@@ -0,0 +1,94 @@
+#include <sputter/net/port.h>
+#include <iostream>
+#include <cstring>
+
+using namespace sputter::net;
+
+// Constants for RX port number, TX port number
+// TODO Have me automatically find a free port
+const int RX_PORT = 50000;
+const int TX_PORT = 50001;
+
+// Example of receiving data:
+// std::string address;
+// uint16_t port;
+
+// int bytesReceived = receivingPort.receive(buffer, 256, address, port);
+
+// if (bytesReceived > 0) {
+//   std::cout << "Received " << bytesReceived << " bytes from " << address << ":" << port << std::endl;
+//   std::cout << "Data: " << buffer << std::endl;
+// }
+
+
+// Client code
+int main(int argc, char* argv[]) {
+  // Parse arguments for remote host
+  char* REMOTE_HOST;
+  if (argc > 1) {
+    REMOTE_HOST = argv[1];
+  } else {
+    std::cout << "Usage: " << argv[0] << " [remote host]" << std::endl;
+    return 1;
+  }
+
+  // Open the client port for receiving data (hole punch)
+  sputter::net::UDPPort receivingPort;
+
+  if (!receivingPort.open(RX_PORT)) {
+    std::cerr << "Error: Failed to open client port\n";
+    return 1;
+  }
+
+  // Send data to the server
+  const char *message = "hole-punch";
+  if (!receivingPort.send(message, strlen(message) + 1, REMOTE_HOST, TX_PORT)) {
+    std::cerr << "Error: Failed to send data\n";
+    return 1;
+  }
+
+  // Open the client port for sending data
+  sputter::net::UDPPort sendingPort;
+
+  if (!sendingPort.open(TX_PORT)) {
+    std::cerr << "Error: Failed to open client port\n";
+    return 1;
+  }
+
+  // Send data until the server sends a message back
+  // First, send a message to the server using the sending port
+  // Then, receive a message from the server using the receiving port
+  // Repeat until a message from the server is received.
+  char buffer[256];
+  std::string address;
+  uint16_t port;
+
+  while (true) {
+    // Send a message to the server
+    const char *message = "Hello, server!";
+    if (!sendingPort.send(message, strlen(message) + 1, REMOTE_HOST, RX_PORT)) {
+      std::cerr << "Error: Failed to send data\n";
+      return 1;
+    }
+
+    // Receive a message from the server
+    int bytesReceived = receivingPort.receive(buffer, 256, address, port);
```

Comment:
This is a solid start, though we're gonna want a nonblocking `receive()` call to facilitate having a "netcode tick phase" in the main game loop.

Reply:
Definitely, wouldn't want to stop simulation while waiting for a peer. 

From my research, even if the socket was non-blocking, we'd need to have a `while()` loop to collect all the data, effectively blocking the thread calling `receive()`. 

If my assumption is correct, do you think a separate thread is required here?

Reply:
For lockstep simulation, a nonblocking + polling model works best I think. Threading to compensate risks priority inversion and other synchronisation costs, and you sorta end up duplicating work the kernel already does on your behalf.

The answer to this Q covers it nicely I think:  https://gamedev.stackexchange.com/questions/64248/udp-non-blocking-or-a-separate-thread-for-receiving

Another high-level game loop example from a game networking book I've been reading:
![image](https://user-images.githubusercontent.com/4349167/216840771-61911353-88b3-48ec-9f4f-212fee8f6f34.jpeg)

I'm not speaking too much from experience on this one fwiw, just the mental model I have working in driverland and regurgitating stuff I've read.